In [1]:
from IPython.core.display import HTML
with open ('style.css', 'r') as file:
    css = file.read()
HTML(css)

# Ein Rangier-Problem

<img src="rangierProblem.png">

Die obige Abbildung zeigt ein *Rangier-Problem*: Auf dem Gleis-Abschnitt <strong>A</strong> befinden sich drei Waggons, die wir mit den Ziffern 1, 2, 3 bezeichnen.  Auf dem Gleisabschnitt <strong>B</strong> befindet sich eine Lokomotive, die wir mit der Ziffer 0 bezeichnen.   Ziel ist es, die Waggons in der Reihenfolge 3, 1, 2 auf dem Gleis-Abschnitt <strong>C</strong> abzustellen.  Die Lokomotive soll am Schluss wieder auf den Gleis-Abschnitt <strong>B</strong> zurückfahren.  Die Lokomotive kann die Waggons in beliebiger Reihenfolge an und abkoppeln.  Beim Rangieren ist es erlaubt, dass die Lokomotive gleichzeitig Waggons vorne und hinten anhängt.  Eine Konstellation der Form <tt>[1,2,0,3]</tt>, bei der die Lokomotive zwei Waggons zieht und einen Waggon schiebt, (oder umgekehrt, je nach dem, in welche Richtung sie fährt) ist also möglich.

## Funktionen zum Ausdrucken der Lösung

Die folgenden beiden Funktionen dienen nur dazu, die Lösung etwas lesbarer auszudrucken.

In [2]:
def printState(A, B, C):
    "print a given state"
    print('A: ', list(A), '; B: ', list(B), '; C: ', list(C), sep='')

In [3]:
def printPath(Path):
    "print the given Path"
    print("Solution:\n")
    for (A, B, C) in Path:
        printState(A, B, C)

## Hilfsfunktionen

**Aufgabe 1**: Definieren Sie eine Funktion $\texttt{allTuples}(S)$, so dass für eine Menge $S$ der Aufruf $\texttt{allTuples}(S)$
die Menge aller der Tupel berechnet, die jedes Element von $S$ genau einmal enthalten.  Beispielsweise soll der Aufruf
$$ \texttt{allTuples}(\{1,2,3\}) $$
die folgende Menge als Ergebnis zurück liefern:
$$ \bigl\{(1, 2, 3), (1, 3, 2), (2, 1, 3), (2, 3, 1), (3, 1, 2), (3, 2, 1)\bigr\} $$
**Hinweis**: Versuchen Sie, diese Funktion rekursiv zu definieren.

In [4]:
def allTuples(S):
    '''
    Given a set S this function returns the set of all tuples that contain each element
    of S exactly once.
    '''
    n = len(S)
    if n == 0:
        return { () }
    T = set(S)
    x = T.pop()
    return { t[:k] + (x,) + t[k:] for k in range(n) for t in allTuples(T) }

In [5]:
allTuples({1,2,3})

{(1, 2, 3), (1, 3, 2), (2, 1, 3), (2, 3, 1), (3, 1, 2), (3, 2, 1)}

Die Funktion $\texttt{power}(M)$ berechnet für eine Menge $M$ die Potenz-Menge $2^M$.

In [6]:
def power(M):
    'This function computes the power set of the set M.'
    if M == set():
        return { frozenset() }
    else:
        C  = set(M)  # C is a copy of M as we don't want to change the set M
        x  = C.pop() # pop removes the element x from the set C
        P1 = power(C)
        P2 = { A | {x} for A in P1 }
        return P1 | P2

Die Funktion $\texttt{inverse}(R)$ berechnet das Inverse der Relation $R$.

In [7]:
def inverse(R):
    'return the inverse of the relation R'
    return { (y, x) for (x, y) in R }

Die Funktion $\texttt{myReverse}(L)$ dreht die Reihenfolge der Elemente in dem Tupel $L$ um.
Versuchen Sie eine rekursive Implementierung.

In [8]:
def myReverse(L):
    'return the reverse of the tuple L'
    n = len(L)
    if n == 0:
        return ()
    return (L[-1],) + myReverse(L[:-1])

In [9]:
myReverse((1,2,3))

(3, 2, 1)

In [10]:
%run Breadth-First-Search.ipynb

## Problemspezifischer Code

Wir stellen die Waggons durch die Ziffern 1, 2 und 3 dar, die Lokomotive wird durch 0 dargestellt.  Zustände werden durch Tupel dargestellt.  Beispielsweise wird der Start-Zustand durch das Tupel
$$ \bigl\langle\langle 1,2,3 \rangle, \langle 0 \rangle, \langle\rangle\bigr\rangle $$
dargestellt, denn auf dem Gleis <b>A</b> stehen die Waggons 1, 2, 3, die Lokomotive steht auf dem Gleis <b>B</b> und das Gleis <b>C</b> ist leer.

In [11]:
All = { 0, 1, 2, 3 }




In [12]:
PowerAll = power(All)
PowerAll

{frozenset(),
 frozenset({2}),
 frozenset({3}),
 frozenset({2, 3}),
 frozenset({0, 2, 3}),
 frozenset({1}),
 frozenset({1, 2}),
 frozenset({1, 3}),
 frozenset({1, 2, 3}),
 frozenset({0}),
 frozenset({0, 1}),
 frozenset({0, 2}),
 frozenset({0, 1, 2}),
 frozenset({0, 3}),
 frozenset({0, 1, 3}),
 frozenset({0, 1, 2, 3})}

**Aufgabe 2**: Definieren Sie die Menge $\texttt{Partitions}$ so, dass diese Menge alle Tripel der Form
$$ \langle A, B, C \rangle $$
enthält, für welche die Menge $\{ A, B, C \}$ eine <em style="color:blue">Partition</em> der Menge $\{0,1,2,3\}$ ist.
<ol> 
<li><b>Hinweis</b>: Die Menge $\{ A, B, C \}$ ist genau dann eine Partition einer Menge $S$, 
    wenn $A$, $B$ und $C$ Mengen sind, so dass gilt:
    $$A \cup B \cup C = S \quad\mbox{und}\quad 
      A \cap B = \{\}, \quad A \cap C = \{\} \quad \mbox{und} \quad B \cap C = \{\}.
    $$  
    </li>
<li><b>Hinweis</b>:  Die Menge $$\{0,1,2,3\} \quad \mbox{hat 81 Partitionen der Form}\quad \langle A, B, C \rangle.$$
    </li>
<ol>

In [13]:
Partitions = {  (A,B,C) for A in PowerAll for B in PowerAll for C in PowerAll
                if A | B | C == All and A & B == set() and A & C == set() and B&C == set()
                                  
             }

**Aufgabe 3**: Wir stellen Zustände durch Tupel der Form
$$ \langle LA, LB, LC \rangle $$
dar.  Dabei ist $LA$ das Tupel der Waggons auf dem Gleis A, $LB$ ist das Tupel der
Waggons auf dem Gleis $B$ und $LC$ ist das Tupel der Waggons auf dem Gleis C.
Berechnen Sie die Menge aller Zustände.

In [15]:
States = { (A,B,C) for (A,B,C) in Partitions
                        for A in allTuples(A)
                        for B in allTuples(B)
                        for C in allTuples(C)}
len(States)
States





{((), (), (0, 1, 2, 3)),
 ((), (), (0, 1, 3, 2)),
 ((), (), (0, 2, 1, 3)),
 ((), (), (0, 2, 3, 1)),
 ((), (), (0, 3, 1, 2)),
 ((), (), (0, 3, 2, 1)),
 ((), (), (1, 0, 2, 3)),
 ((), (), (1, 0, 3, 2)),
 ((), (), (1, 2, 0, 3)),
 ((), (), (1, 2, 3, 0)),
 ((), (), (1, 3, 0, 2)),
 ((), (), (1, 3, 2, 0)),
 ((), (), (2, 0, 1, 3)),
 ((), (), (2, 0, 3, 1)),
 ((), (), (2, 1, 0, 3)),
 ((), (), (2, 1, 3, 0)),
 ((), (), (2, 3, 0, 1)),
 ((), (), (2, 3, 1, 0)),
 ((), (), (3, 0, 1, 2)),
 ((), (), (3, 0, 2, 1)),
 ((), (), (3, 1, 0, 2)),
 ((), (), (3, 1, 2, 0)),
 ((), (), (3, 2, 0, 1)),
 ((), (), (3, 2, 1, 0)),
 ((), (0,), (1, 2, 3)),
 ((), (0,), (1, 3, 2)),
 ((), (0,), (2, 1, 3)),
 ((), (0,), (2, 3, 1)),
 ((), (0,), (3, 1, 2)),
 ((), (0,), (3, 2, 1)),
 ((), (0, 1), (2, 3)),
 ((), (0, 1), (3, 2)),
 ((), (0, 1, 2), (3,)),
 ((), (0, 1, 2, 3), ()),
 ((), (0, 1, 3), (2,)),
 ((), (0, 1, 3, 2), ()),
 ((), (0, 2), (1, 3)),
 ((), (0, 2), (3, 1)),
 ((), (0, 2, 1), (3,)),
 ((), (0, 2, 1, 3), ()),
 ((), (0, 2, 3), 

**Hinweis**:  Es gibt 360 verschiedene Zustände.

In [16]:
len(States)


360

**Aufgabe 4**: Berechnen Sie die Menge aller Transitionen, in denen die Lokomotive vom Gleis <b>A</b> über das westliche Gleis zum Gleis <b>C</b> fährt. 

<b>Hinweis</b>: Die Funktion $\texttt{myReverse}$ dreht ein Tupel um.

In [17]:
RacWest = { ((a,b,c),(a[i:],b,myReverse(a[:i])+c)) for (a,b,c) in States 
                                                    for i in range(1,len(a)+1) 
                                                    if 0 in a[0:i]
}

len(RacWest)

210

**Hinweis**: Es gibt 210 verschiedene Transitionen in der Menge `RacEast`.

**Aufgabe 5**: Berechnen Sie die Menge aller Transitionen, in denen die Lokomotive vom
Gleis <b>A</b> über das östliche Gleis zum Gleis <b>C</b> fährt.

In [18]:
RacEast = { ((a,b,c),(a[:i],b,c+myReverse(a[i:]))) for (a,b,c) in States 
                                                    for i in range(0,len(a)) 
                                                    if 0 in a[i:]
}
len(RacEast)

210

Auch diese Menge enthält 210 Elemente.

In [19]:
Rac = RacEast | RacWest

len(Rac)


360

In [20]:
Rca = inverse(Rac)



**Aufgabe 6**:
Berechnen Sie die Menge aller Transitionen, in denen die Lokomotive vom Gleis <b>B</b> zum Gleis <b>C</b> fährt.

In [21]:
Rbc = { ((a,b,c),(a,b[:i],b[i:]+c)) for (a,b,c) in States 
                                    for i in range(0, len(b) )
                                    if 0 in b[i:]}
           
      
         

Die Menge <tt>Rbc</tt> enthält 210 Elemente. 

In [22]:
len(Rbc)

210

In [23]:
Rcb = inverse(Rbc)

len(Rcb)

210

In [24]:
R = Rac | Rca | Rbc| Rcb


Die Relation $R$ enthält 1140 verschiedene Elemente.

In [25]:
len(R)

1140

In [26]:
%%time
start = ((1, 2, 3), (0,), ())
goal  = ((), (0,), (3, 1, 2))
Path  = search(R, start, goal)

Iteration 1: 1 new states found.
Iteration 2: 2 new states found.
Iteration 3: 6 new states found.
Iteration 4: 20 new states found.
Iteration 5: 9 new states found.
Iteration 6: 20 new states found.
Iteration 7: 18 new states found.
Iteration 8: 34 new states found.
Iteration 9: 26 new states found.
Iteration 10: 51 new states found.
Iteration 11: 42 new states found.
Wall time: 22.9 ms


The tuple **<tt>Path</tt>** that is a solution to the shunting problem has a length of **13**.  If your path is shorter, then you have to inspect it **carefully** to identify the problem.

In [27]:
len(Path)

13

In [28]:
printPath(Path)

Solution:

A: [1, 2, 3]; B: [0]; C: []
A: [1, 2, 3]; B: []; C: [0]
A: [0, 1, 2, 3]; B: []; C: []
A: [2, 3]; B: []; C: [1, 0]
A: [2, 3]; B: [1, 0]; C: []
A: [2, 3]; B: [1]; C: [0]
A: [0, 2, 3]; B: [1]; C: []
A: [3]; B: [1]; C: [2, 0]
A: [3]; B: [1, 2, 0]; C: []
A: [3]; B: []; C: [1, 2, 0]
A: [3, 0]; B: []; C: [1, 2]
A: []; B: []; C: [0, 3, 1, 2]
A: []; B: [0]; C: [3, 1, 2]


In [29]:
#yayyyay